### DISCUSSION TITLE
expired orm ignores `expunge` and tries to refresh orm on attribute access

### DISCUSSION THREAD

**Discussion by valq7711** [sqlalchemy/sqlalchemy#12052] [Usage Questions] [answered]
SQLAlchemy 1.4.54

```python
from sqlalchemy import Column, Integer, String, select
from sqlalchemy.orm import declarative_base
from sqlalchemy import create_engine
from sqlalchemy.orm import Session


Base = declarative_base()


class User(Base):
    __tablename__ = "user_account"
    id = Column(Integer, primary_key=True)
    name = Column(String(30))

    def __repr__(self):
        return f"User(id={self.id!r}, name={self.name!r})"


engine = create_engine("sqlite://", echo=True, future=True)

Base.metadata.create_all(engine)


with Session(engine) as session:
    session.add(User(id=1, name='Tom'))
    session.commit()


with Session(engine) as session:
    q = select(User).limit(1)
    orm = session.execute(q).scalar_one()
    print(orm)
    print(orm.id)
    session.commit()
    session.expunge(orm)

print(orm.id)  # ERROR sqlalchemy.orm.exc.DetachedInstanceError: Instance <User at 0x289e9e63ee0> is not bound to a Session; ...
```
I know about `expire_on_commit=False` and if I put `expunge` before `commit` it works.
The issue is about general logic - detached orm should behave as detached

**Comment by zzzeek** [comment]
seems to be working as expected

"orm" is detached and expired by the .commit().  so...it's got nothing in it, can't load new data.

**Reply by valq7711** [reply]
The problem is that it **tries** to load while in **detached** state. Shouldn't it stop tracking attribute access in detached state (after expunge-call)?

**Reply by zzzeek** [reply]
well you are asking it what the value of `orm.id` is, and it's gone.  so it can't tell you the answer

**Reply by valq7711** [reply]
Do you mean that after expiration the orm-instance becomes empty? I expected it to contain the last state - just before expiration

**Reply by zzzeek** [reply, marked as answer]
expire erases what was there

in this specific case, *technically* we do have that data because it's the PK and it's inside the state, but when accessing from the ORM mapped attribute we treat this like any other expired value

if you are working with detached objects at all I would strongly advise turning expire_on_commit off
